### DISCUSSION TITLE
MOC-129823 tictactoe running pybricks python

### DISCUSSION THREAD

**Discussion by BertLindeman** [pybricks/support#1219] [Show and tell]
Hello,

Rewrote the program for [Rebrickable moc 129823 tictactoe](https://rebrickable.com/mocs/MOC-129823/3Lego1Math/robotic-arm-tictactoe/#details) from LEGO Spike-prime python into pybricks.

Robot design, instructions and SpikePrime python program from [3Lego1Math](https://rebrickable.com/users/3Lego1Math/)
His movie is way better than mine 👍 

There are some minor modifications to allow calibration, like a left and right stop for the house rotation.

The studio-file and the pdf are on my  [bricksafe directory](https://bricksafe.com/pages/Bert_bricksafe/moc-129823-tictactoe/building-instructions)

The zip with the programs on [bricksafe](https://bricksafe.com/pages/Bert_bricksafe/moc-129823-tictactoe/pybricks-programs)

Read the readme first, so you know to run the calibration program first and paste (some of the) output into the config program.
But, better read it yourself 😃 

Want to see a movie first?
- a bit dull movie of [part of the calibration](https://www.youtube.com/watch?v=0NX1Wp10hXA) (also included in the other movie):
- A [computer to human tictactoe play](https://www.youtube.com/watch?v=Zjy9SLmatk8)

Enjoy,

Bert

**Comment by dlech** [comment]
Really nice! The motion is nice and smooth. No big bounce at the end of a maneuver like I've seen in a lot of other robotic arms.

Pro tip of the day `micropython.const()` only has an effect on uses of the identifier in the file where it is used. So importing something "const" from another file is no different at all from importing something without const. So in `tictacktoe_config.py` it isn't doing anything at all.

**Reply by BertLindeman** [reply]
Thank you David. Let those pro tips come and come. 

Tips from Laurens about torque and dutylimit helped a lot.

I thought it would look sofisticated to use the const. Oh, well better next time. 
Would it make a performance or space difference  if used as it's supposed to be?

**Reply by dlech** [reply]
Yes, when you use an name defined with const in the same file, the compiler will substitute the value which avoids the name look up. This is a very small performance improvement. The difference probably won't be noticeable in most Pybricks program unless you are writing something that is very CPU intensive.

If you start the name of a constant with a leading `_`, then a string containing that name won't be saved in the binary .mpy file. This can be a nice space savings, particularly if you are trying to cram a big program on to a move hub.

example1.py

```python
_MY_CONST = const(10)

...

# might actually speed things up here since it is evaluated 1000 times in a row
values = [x * _MY_CONST for x in range(1000)]

# compared to

_NOT_CONST = 10
values = [x * _NOT_CONST for x in range(1000)]
```

example2.py

```python
# this will raise an error because the name '_MY_CONST' was optimized out by the compiler
from example1 import  _MY_CONST
```
